### PR TITLE
feat(mysql): support `order by nulls first/last`

### LINKED ISSUE #5004: MySQL Order By `*_NULLS_(LAST|FIRST)` incorrect syntax
### Describe the bug

When using `orderBy` in a find operation, using `ASC_NULLS_LAST` or equivalent will insert the syntax `order by ... asc nulls last`. This is incorrect syntax.

If we have a table

```sql
CREATE TABLE foo
(
  id            INT UNSIGNED AUTO_INCREMENT
    PRIMARY KEY,
  text          VARCHAR(255) NOT NULL,
  optional_text VARCHAR(255) NULL,
);
```

Then to sort nulls on `optional_text`, we need to do the following:

| Operation        | Syntax                                                   |
|------------------|----------------------------------------------------------|
| ASC NULLS FIRST  | `order by optional_text is not null, optional_text`      |
| ASC NULLS LAST   | `order by optional_text is null, optional_text`          |
| DESC NULLS FIRST | `order by optional_text is not null, optional_text desc` |
| DESC NULLS LAST  | `order by optional_text is null, optional_text desc`     |

### Reproduction

https://github.com/kpervin/MikroORM-Issue-Example-Repo/blob/c446d1fd0b313da3c1ccb97915fc6a2fd09003ae/test/mysql-order-by-issue.spec.ts#L49-L73

### What driver are you using?

@mikro-orm/mysql

### MikroORM version

5.9.4

### Node.js version

20.9.0

### Operating system

Ubuntu 20.04 LTS

### Validations

- [X] Read the [Contributing Guidelines](https://github.com/mikro-orm/mikro-orm/blob/master/CONTRIBUTING.md).
- [X] Read the [docs](https://mikro-orm.io/docs).
- [X] Check that there isn't [already an issue](https://github.com/mikro-orm/mikro-orm/issues) that reports the same bug to avoid creating a duplicate.
- [X] Check that this is a concrete bug. For Q&A open a [GitHub Discussion](https://github.com/mikro-orm/mikro-orm/discussions) or join our [Slack](https://join.slack.com/t/mikroorm/shared_invite/enQtNTM1ODYzMzM4MDk3LWM4ZDExMjU5ZDhmNjA2MmM3MWMwZmExNjhhNDdiYTMwNWM0MGY5ZTE3ZjkyZTMzOWExNDgyYmMzNDE1NDI5NjA).
- [X] The provided reproduction is a [minimal reproducible example](https://stackoverflow.com/help/minimal-reproducible-example) of the bug.

### PR DESCRIPTION
**Pull request by kpervin**
Fixes #5004

- Added `getOrderByExpression` to AbstractSqlPlatform.ts as an overridable method to format `orderBy` queries. This uses syntax that is recognized by both SQLite and PostGreSQL
- Added override of `getOrderByExpression` to MariaDbPlatform.ts and MySqlPlatform.ts that adds functionality to handle `(ASC|DESC) NULLS (FIRST|LAST)` syntax

### CODE REVIEW

File: packages/mariadb/src/MariaDbPlatform.ts [hunk 1]
```diff
@@ -74,4 +74,21 @@ export class MariaDbPlatform extends AbstractSqlPlatform {
...
   }
 
+  private readonly NULLS_TRANSLATE = {
+    [QueryOrder.asc_nulls_first]: 'is not null',
+    [QueryOrder.asc_nulls_last]: 'is null',
+    [QueryOrder.desc_nulls_first]: 'is not null',
+    [QueryOrder.desc_nulls_last]: 'is null',
+  } as const;
+
+  override getOrderByExpression(column: string, direction: string): string[] {
```
Comment (B4nan) [inline packages/mariadb/src/MariaDbPlatform.ts:84]: i guess we dont need to test the mariadb explicitly, lets just add `/* istanbul ignore next */` here
```suggestion
  /* istanbul ignore next */
  override getOrderByExpression(column: string, direction: string): string[] {
```

**Review by B4nan — COMMENTED** [review]
can you check how sqlite behaves? we might need the same there

> This uses syntax that is recognized by both SQLite and PostGreSQL

edit: i should read the whole thing first :]

**Comment by kpervin** [comment]
> can you check how sqlite behaves? we might need the same there
> 
> > This uses syntax that is recognized by both SQLite and PostGreSQL
> 
> edit: i should read the whole thing first :]

It probably would make sense to create test suites for each language. Currently the QueryBuilder.test.ts file is only testing against MySQL from what I could see.

File: packages/mariadb/src/MariaDbPlatform.ts [hunk 2]
```diff
@@ -74,4 +74,22 @@ export class MariaDbPlatform extends AbstractSqlPlatform {
     return `alter table ${quotedTableName} add fulltext index ${quotedIndexName}(${quotedColumnNames.join(',')})`;
   }
 
+  private readonly NULLS_TRANSLATE = {
```
Comment (B4nan) [inline packages/mariadb/src/MariaDbPlatform.ts:77]: maybe lets remove those to `ORDER_BY_NULLS_TRANSLATE`?

File: packages/mariadb/src/MariaDbPlatform.ts [hunk 3]
```diff
@@ -74,4 +74,22 @@ export class MariaDbPlatform extends AbstractSqlPlatform {
...
+
+  /* istanbul ignore next */
+  override getOrderByExpression(column: string, direction: string): string[] {
+    const ret: string[] = [];
+    const dir = direction.toLowerCase();
+    if (dir in this.NULLS_TRANSLATE) {
+      ret.push(`${column} ${this.NULLS_TRANSLATE[dir as keyof typeof this.NULLS_TRANSLATE]}`);
+    }
+    ret.push(`${column} ${dir.replace(/(\s|nulls|first|last)*/gi, '')}`);
+    return ret;
```
Comment (B4nan) [inline packages/mariadb/src/MariaDbPlatform.ts:92]:
```suggestion
    const dir = direction.toLowerCase() as keyof typeof this.NULLS_TRANSLATE;

    if (dir in this.NULLS_TRANSLATE) {
      ret.push(`${column} ${this.NULLS_TRANSLATE[dir]}`);
    }

    ret.push(`${column} ${dir.replace(/(\s|nulls|first|last)*/gi, '')}`);

    return ret;
```

**Review by B4nan — COMMENTED** [review]
just two nits, please apply those suggestions to both drivers

otherwise ready to merge, thanks!

> It probably would make sense to create test suites for each language. Currently the QueryBuilder.test.ts file is only testing against MySQL from what I could see.

part of that file is also about postgres, but agreed this would deserve a split and other drivers, probably not a complete copy of what we have there but at least for the differences

File: packages/mariadb/src/MariaDbPlatform.ts (hunk 2 above)
Comment (kpervin) [reply packages/mariadb/src/MariaDbPlatform.ts:77]: Sorry, did you mean rename (not remove)?

Comment (B4nan) [reply packages/mariadb/src/MariaDbPlatform.ts:77]: yes

**Review by B4nan — APPROVED** [review]

**Pull request merged**